### DISCUSSION TITLE
Compatibility with future keycloak version and security considerations

### DISCUSSION THREAD

**Discussion by jnachtigall** [keycloakify/keycloakify#346] [General]
The selling points of Keycloakify are pretty obvious - especially if one uses React anyway for the frontend - and well explained on https://www.keycloakify.dev/ 

While evaluation theming solutions for Keycloak I came across a few issues. Some these issues might be considered downsides/cons. Maybe those point could be clarified in the documentation or website. I hope I can give some feedback here.

**1. Support for future keycloak version?**

Obviously Keycloakify has more or less one maintainer. What happens if keycloakify gets unmaintained while Keycloak keeps releasing new versions. How much work would it be to keep my keycloakify based theme then? 

Also https://docs.keycloakify.dev/readme-1#see-versions-keycloakify-have-been-tested-with says
> However, the default pages you will get (before you customize them) will always be the ones of Keycloak v11.03 and some extra pages that didn't existed back then like register-user-profile.ftl.  

Version 11 is pretty old. I always try to use the most recent version which is 21.1.1 So this gives the impression as if Keycloakify would be based on a pretty old Keycloak version?

I am not sure if the documentation is misleading here and could be improved and clarified (that my main reason for opening this issue/discussion).

**2. Security considerations**

Related to this the question might be raised what happens if there is, for instance a (security) hot fix release, for the official Keycloak theme. If I base my work on the official Keycloak Theming then will it be "more secure" than using Keycloakify (which is based on Keycloak v11 and would be lots of work in case it becomes unmaintained.)

I hope my points are well received. I am honestly not sure if they are really valid. Maybe the _documentation_ could be improved regarding **maintenance work**, **future compatibility** or **security considerations** regarding keycloakify based themes?

**Comment by garronej** [comment]
Hi @jnachtigall,

Thank you for reaching out! Your questions are perfectly valid, and I am happy to provide more insight.

> Obviously Keycloakify has more or less one maintainer.

This isn't entirely clear. @lordvlad has been consistently contributing for some time now. Additionally, @willwill96 has shown to be quite involved in the project and demonstrates an in-depth understanding.

> What happens if keycloakify gets unmaintained while Keycloak keeps releasing new versions? How much work would it be for us to keep our keycloakify based theme then?

Even if I were to stop maintaining Keycloakify (which I don't plan on doing), it's highly unlikely that Keycloak would release a new version that breaks Keycloakify themes. This would mean that they would break every other theme, whether Keycloakify-based or not. However, this is merely speculation. What I can say for certain is that I'm firmly committed to maintaining and growing Keycloakify for the foreseeable future.

> Version 11 is pretty old. We always try to use the most recent version, which is 21.1.1. So, this gives the impression that Keycloakify would be based on a pretty old Keycloak version?

Yes, this part of the documentation is outdated and can be misleading. We ensure that Keycloakify generates themes compatible with the latest Keycloak version. We use Keycloakify as a side project of [Onyxia Datalab](https://onyxia.sh) and always maintain up-to-date Keycloak versions. If anything goes wrong after upgrading Keycloak, we'll be among the first to know.

In fact, Keycloakify consists of two things:
- A build tool that compiles Keycloak themes.
- A set of React components used as a starting point for building your custom theme, showcased in [the storybook](https://storybook.keycloakify.dev). These components use the resources (e.g., PatternFly, background image, etc. [check it out](https://github.com/keycloak/keycloak/tree/11.0.3/themes/src/main/resources/theme/keycloak/common/resources)) provided with Keycloak v11, and we haven't bothered to update these assets since most people use Keycloakify to replace them. If it's a concern, we could consider updating these assets to the latest version periodically. Now that we have visual regression testing, it should be easy to implement. We haven't done it yet because no one seemed to care enough to open an issue about it.

> Related to this, the question was raised about what happens if there is, for instance, a (security) hotfix release for the official Keycloak theme. If we base our work on the official Keycloak Theming, then we are "more secure" than using Keycloakify (which is based on Keycloak v11 and would require a lot of work in case it becomes unmaintained).

The Keycloakify community and I are very prompt in reacting to security vulnerabilities. [It happened once, and we had a patch released before Keycloak did](https://github.com/keycloakify/keycloakify/issues/217). To be fully honest, the vulnerability was on the Keycloak side, but it wasn't exploitable with the default theme. If there is a theme-related security patch on the Keycloak side, we'll be sure to apply it to Keycloakify swiftly.

> I hope my points are well received.

They certainly are! Thank you for your valuable feedback. I'll make sure to update the documentation accordingly.

**Comment by jnachtigall** [comment]
Hi @garronej,

thanks for the answer. Much appreciated. I hope things can better explained in the documentation so it easier to understand.

> In fact, Keycloakify consists of two things.

Good point. Related to this is another point that made it hard for me to estimate the afford needed to keep a keycloakify based theme up to date (if in case the keycloakify upstream projects would go unmaintained): How does a Keycloakify/React based theme work technically under the hood? That is, how does the communication work between Keycloakify server and the Theme? I looked into the files (mainly `*.ftl` markup files) that Keycloakify generates. And it seemed like Keycloakify generates generic `*.ftl` pages like a "normal" theme. And these `*.ftl` files includes the React App (based on Create-React-App). But how do the React app and the Theme exchange data? – afaics there is some generated JSON (how? with `kcContext.ts` or similar) within the ftl file that gets is used as input to the React App. But trying to understand how Keycloakify works (and I am also new to Keycloak theming in general - I rather have a React background) – so looking into the technical details felt pretty much like reverse engineering the built dist files. And not fully understanding the whole process here what it hard to understand how deeply coupled Keycloakify is to a distinct build/version of Keycloak itself. Is it just generation of JSON using a stable Keacloakify API (loose coupling => good), does it hook into internal Keycloak functions (this might break when upgrading keycloak => bad)?

So a chapter like "How does Keycloakify work technically?" (or maybe rather "How does the generated React theme work with regards to Keycloak backend server?") in the documentation would really have helped me.

FWIW, I did not use the CSS only theming approach while evaluating but rather my own React components (which are based on MUI but that's different for each project). That is, the using https://github.com/keycloakify/keycloakify-starter/ version 3 (it was some weeks ago)

**Comment by garronej** [comment]
Let me clarify how Keycloak and JavaScript communicate in a more comprehensive manner:

FreeMarker (.ftl) is a templating language, similar to PHP. Keycloak processes the template whenever a request is made. For instance, [here](https://github.com/keycloak/keycloak/blob/04ab84800312a705e3ee623f64fbcc2e2abf078b/themes/src/main/resources/theme/base/login/login.ftl#L9), you can see a global `url` object with a `loginAction` property.

Keycloakify moves the page generation from the backend to the frontend, making the FTL context available to JavaScript. A simple way to achieve this is shown in the [keycloak-theme-vuejs project](https://github.com/wrsouza/keycloak-theme-vuejs/blob/master/src/static/template.ftl), where all known FTL variables are manually listed page by page and transformed into a global JavaScript object. The core Keycloak theme also [implements this strategy](https://github.com/keycloak/keycloak/blob/04ab84800312a705e3ee623f64fbcc2e2abf078b/themes/src/main/resources/theme/keycloak.v2/account/index.ftl#L12-L39) for their latest React-based theme.

However, this approach requires a lot of manual work and is tightly coupled to a specific Keycloak version. If Keycloak introduces a new variable, it won't be visible unless manually added. Additionally, custom Keycloak plugins adding custom variables may cause issues, and if a variable is removed, the template may crash.

Keycloakify takes a more generic approach by using a generic FTL function that converts FTL objects into JSON format. It passes the `.data_model` (FTL equivalent of `window`) to generate a `window.kcContext` representing the FTL context. This method is powerful, future-proof, and the FTL function cannot throw an exception. Every potentially throwable instruction is surrounded by `attempt/recover` (FTL equivalent of `try/catch`).

[Here](https://github.com/keycloakify/keycloakify/blob/0c65561bcbe1c5c05472d72a3f69773cd064bb45/src/bin/keycloakify/generateFtl/ftl_object_to_js_code_declaring_an_object.ftl#L162-L196), you can see some context-specific code. This is to avoid log pollution when errors occur, even when using `attempt/recover`. Special cases are made for variables known to throw errors on specific pages. For instance, [here](https://github.com/keycloakify/keycloakify/blob/0c65561bcbe1c5c05472d72a3f69773cd064bb45/src/bin/keycloakify/generateFtl/ftl_object_to_js_code_declaring_an_object.ftl#L189-L191C12), `realm.name` is not listed in `error.ftl` because it would log an error. Keycloakify has [a documentation page](https://docs.keycloakify.dev/keycloak-error-in-log) for users to report log errors, which are then addressed with special cases.

Regarding functions, [specific handling](https://github.com/keycloakify/keycloakify/blob/0c65561bcbe1c5c05472d72a3f69773cd064bb45/src/bin/keycloakify/generateFtl/ftl_object_to_js_code_declaring_an_object.ftl#L31-L110) is necessary since we cannot invoke functions without knowing their required parameters ahead of time. However, the only functions present are well-known ones like `printIfExists`, `existsError`, etc.

Keycloakify also implements strategies for bundling assets and ensuring they are served and resolved correctly within Keycloak. While you don't need to understand the precise workings to make a decision, you can refer to @willwill96's video for some insights: https://youtu.be/x3ux2JM1Bxk

To answer your question about whether Keycloakify relies on a stable API or internal Keycloak functions: Keycloakify doesn't depend on any Keycloak API. Instead, it translates the FTL context into a JavaScript object (kcContext), making minimal assumptions about the FTL context.

This explanation should help clarify things and reassure you that Keycloakify is relatively future-proof regarding upcoming Keycloak updates, even if maintenance were to cease.

**Reply by shighetari** [reply]
Very insightful read, thank you for the information and references.

**Reply by garronej** [reply]
UPDATE: Oct 2025. Keycloakify nows automatically serializing classes that have no buisness being in the FreeMarker context.  
See: https://github.com/keycloakify/keycloakify/pull/907#issuecomment-3413939448
